src/webview/js/js.js

```diff
@@ -511,6 +511,19 @@ const scrollToPreserve = (msgId: number, prevBoundTop: number) => {
   window.scrollBy(0, newBoundRect.top - prevBoundTop);
 };
 
+// Run a function after the next repaint. This is important if we have just set
+// innerHTML and need to read from the DOM, as the repaint happens asynchronously.
+// See: https://macarthur.me/posts/when-dom-updates-appear-to-be-asynchronous
+const runAfterRepaint = (fn: () => void) => {
+  requestAnimationFrame(() => {
+    // this runs before the next repaint
+    requestAnimationFrame(() => {
+      // this runs after the repaint
+      fn();
```

Comment:
So one thing that's less than optimal about this is that it means we're guaranteed to have a frame appear with an intermediate state: we've introduced the new content, but we're at some uncontrolled scroll position that may not be the one we want.

That's still a huge improvement over the bug this fixes -- better to be wrong for a frame than to be wrong and stay that way.

This seems likely to be the cause of the jank this introduces on Android, described above. I imagine it might also cause some jank on iOS, too.

Reply:
One thing we could try instead is `setTimeout(fn, 0)` inside the first rAF. In theory this could run a bit earlier but I don't have a good idea if that'll be less reliable.

Reply:
Yeah. And I think it'll still result in showing a frame based on the intermediate state -- AIUI the point of rAF is that its callback gets invoked just at the last moment before the frame gets turned to pixels.

What about just one rAF, and invoking the callback directly in that? Not sure if you've already tested and found that `scrollHeight` isn't yet updated at that point, but a priori it seems like it might be. The blog post is doing a different test, of whether what's visible on the screen has been updated -- and it'd actually make a lot of sense for the API to be that when a `requestAnimationFrame` callback is invoked, all the DOM layout properties are guaranteed to be ready and just the repaint hasn't happened yet.

Reply:
Hmm, I think I may have messed something up when testing a single rAF earlier (I probably failed to regenerate the webview JS). It seems to be working with only one call, which I agree makes sense. I'm going to do some more testing but I'm not seeing a difference. The same content update jank is still there, but that's unsurprising.

Reply:
Ah, aha, here's some previous discussion of this: https://github.com/zulip/zulip-mobile/issues/3457#issuecomment-555252129

I *thought* I remembered a discussion of `requestAnimationFrame` in a context like this, but I was having trouble finding it. There it is.

The conclusion there is... that either a `setTimeout` inside the first rAF, or just running the code directly inside a single rAF, are likely solutions :slightly_smiling_face: . Just needs testing to determine which, if either, actually seems to work well.